### PR TITLE
ARROW-4501: Fix out-of-bounds read in DoubleCrcHash

### PR DESCRIPTION
**Pull request by TocarIP**

### CODE REVIEW

**Review by wesm — COMMENTED** [review]
This looks right, but I'll let @pitrou review also. Can you add a test case for the downstream failure in Unique that you found?

File: cpp/src/arrow/util/hashing-test.cc [hunk 1]
```diff
@@ -126,6 +126,19 @@ TEST(HashingQuality, Strings) {
             0.96 * static_cast<double>(2 * values.size()));
 }
 
+TEST(HashingBounds, Strings) {
+  std::string s = "0123456789012345x";
+  hash_t h = ComputeStringHash<1>(&s[0], s.size());
+  int differnt = 0;
+  for (char i = 0; i < 120; i++) {
+      s[16] = i;
```
Comment (pitrou) [inline cpp/src/arrow/util/hashing-test.cc]: Rather than hardcode `16`, should it be `s.size() - 1`?

File: cpp/src/arrow/util/hashing-test.cc [hunk 2]
```diff
@@ -126,6 +126,19 @@ TEST(HashingQuality, Strings) {
...
+      if (ComputeStringHash<1>(&s[0], s.size()) != h) {
+	  differnt++;
+      }
+  }
+  ASSERT_GE(differnt, 100);
```
Comment (pitrou) [inline cpp/src/arrow/util/hashing-test.cc]: Are there that many collisions? I would expect at least 118...

**Review by pitrou — COMMENTED** [review]
Thanks for the fix. Ideally I would like a more general test but we can go with that. Please see comments below, though.

File: cpp/src/arrow/util/hashing-test.cc [hunk 3]
```diff
@@ -126,6 +126,19 @@ TEST(HashingQuality, Strings) {
             0.96 * static_cast<double>(2 * values.size()));
 }
 
+TEST(HashingBounds, Strings) {
+  std::string s = "0123456789012345x";
+  hash_t h = ComputeStringHash<1>(&s[0], s.size());
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/hashing-test.cc]: The more idiomatic is to use `s.c_str()`.

**Review by fsaintjacques — COMMENTED** [review]
I would have expected valgrind/ASAN to catch this.

**Comment by pitrou** [comment]
Not if the memory allocation is larger than the actual string, I suppose.

**Comment by fsaintjacques** [comment]
possibly what happened here, if there's an extra null byte.

**Comment by xhochy** [comment]
Rebased.

**Review by pitrou — APPROVED** [review]
+1. Thank you @TocarIP.

**Pull request closed without merging**